### DISCUSSION TITLE
Can renovate bot use the auth.json within the repository for composer packages?

### DISCUSSION THREAD

**Discussion by amenk** [renovatebot/renovate#22818] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.120.0

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Renovate fetches privat packages from composer by using the auth.json which is commited to the repository.

### What you tried so far.

Checked github issues, looks like it is not suppored.

We could define hostRules, but would like to avoid that and use the data from the repository.

Question: Is renovate using the composer.phar for this or does it have it's own logic?

EDIT: I saw https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/composer/artifacts.ts which seems to inject custom data into the auth.json - the question is, if the existing auth data also could be used.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: GET https://repo.backpackforlaravel.com/packages.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=429 retryCount=0, duration=65) (repository=556)
DEBUG: GET https://packagist.org/p2/backpack/pro.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=114) (repository=556)
DEBUG: Datasource 404 (repository=556)
       "datasource": "packagist",
       "packageName": "backpack/pro",
       "url": "https://packagist.org/p2/backpack/pro.json"
DEBUG: Failed to look up packagist package backpack/pro (repository=556, packageFile=composer.json, dependency=backpack/pro)
```
</details>

### Details

Our composer JSON contains:

```
    "require": {
        "backpack/pro": "^1.0.0",
         ...

    "repositories": [
        {
            "type": "composer",
            "url": "https://repo.backpackforlaravel.com/"
        }

```

And the `auth.json` in the same folder, commited to git:

```
{
    "http-basic": {
        "backpackforlaravel.com": {
            "username": "xxx",
            "password": "yyyy"
        }
    }
}
```

A `composer install` / `composer update` outside of renovate works without problems.

**Comment by rarkins** [comment]
I am not judging, just asking: do you mean that you have hardcoded credentials committed to the repo, which `composer` needs?

In addition, does Renovate find this dependency and fail to look it up due to lack of credentials?

**Reply by amenk** [reply]
Yes and yes :-)

Those credentials are for semi-private packages, such as backpack pro and Magento which are needed for setting up the application. We are talking only about company-internal packages, not public github packages where this auth.json is commited to. The credentials are not used-bound.

I did some more digging and think https://github.com/renovatebot/renovate/blob/4aba6a90ed47430ff8289526f7ee6422b1ba62d3/lib/modules/manager/composer/artifacts.ts#L39 should load an eventually existing `auth.json` instead of starting with an empty one.

**Reply by rarkins** [reply]
Yes, if it was initialized from `auth.json` it could work. It is not written back to `auth.json` and is instead exported in env (which takes higher precedence than disk)

**Reply by amenk** [reply]
Looks like in the first run, renovate native javascript code is accessing the package files, not composer. So the auth.json would need to be actually parsed and passed to renovate ...

In addition, I run quite quick into HTTP 429 rate limit errors from backpacks' cloudflare proxy after the initial 404 error on the packages.

**Reply by rarkins** [reply]
In that case what you likely need is for the `composer` manager in Renovate to be able to detect/parse `auth.json` during the extract phase. Until that's supported, you would need to define hostRules

**Reply by amenk** [reply]
Yes, wondering if other managers have something similar we can build upon.
There is https://docs.renovatebot.com/self-hosted-configuration/#detectglobalmanagerconfig
But in our use case it would be like "detectLocalManagerConfig", because this is per repository, not for the full renovate run, if we target multiple projects in one run.

**Reply by rarkins** [reply]
It would be part of extract, and not need a new config option. The only thing similar is npmrc files.

**Reply by amenk** [reply]
I did some more research and think we would have to switch to extractAllPackageFiles to be able to also include the `auth.json` in the extract process? Or at least we have to add `auth.json` to the [fileMatch](https://github.com/renovatebot/renovate/blob/cf7c0c9a42aa74dbf24464695a5e450e6071f809/lib/modules/manager/composer/index.ts#L21-L21) for composer. Then we could pass the credentials down and use them in [getHostOpts](https://github.com/renovatebot/renovate/blob/cf7c0c9a42aa74dbf24464695a5e450e6071f809/lib/modules/datasource/packagist/index.ts#L35-L35) 

It is also possible to store credentials directly in the [composer.json file in a config node](https://getcomposer.org/doc/articles/authentication-for-private-packages.md#authentication-in-composer-json-file-itself). I believe this is also not yet supported and could be implemented with similar parsing code.

**Reply by rarkins** [reply]
Adding to fileMatch is not the best idea.

If the file contains only credentials and those those credentials only need detecting once, then you don't need to extract all files either - you just need the composer.json to check for a sibling auth.json and return hostRules if found

**Reply by amenk** [reply]
Yes, I was thinking we can look for a `auth.json` in the same directory as the `composer.json`, first. But the code looks decoupled in a way, were (I think), the path to the `composer.json` is not available in [packagist/index.ts](https://github.com/renovatebot/renovate/blob/cf7c0c9a42aa74dbf24464695a5e450e6071f809/lib/modules/datasource/packagist/index.ts#L35-L35) or [extractPackageFile](https://github.com/renovatebot/renovate/blob/cf7c0c9a42aa74dbf24464695a5e450e6071f809/lib/modules/manager/composer/extract.ts#L5-L5).


And should we go on with the discussion here or should a feature request be created under issues?

**Reply by rarkins** [reply]
Feature requests are created by maintainers only, and only once a firm set of requirements are available. This still at the "what if we?" stage

**Reply by amenk** [reply]
I did some tests and I think this is not working with a bearer (for example for the Shopware store package repository)
Can it be considered as a feature requests or was eventually something like this already implemented?

**Comment by amenk** [comment, marked as answer]
I actually found a workaround which would work for us pretty well. Instead of using the auth.json we can include the credentials directly into the repository URL in the composer.json:

```
            "url": "https://user:password@repo.backpackforlaravel.com/"
```

**Reply by amenk** [reply]
Wondering if that would also work with a Bearer somehow, but don't know where to put it in the `composer.json` and if and which of the options renovate would support :thinking:
